rcl/src/rcl/event.c

```diff
@@ -79,10 +69,29 @@ rcl_publisher_event_init(
       RCL_SET_ERROR_MSG("Event type for publisher not supported");
       return RCL_RET_INVALID_ARGUMENT;
   }
-  return rmw_publisher_event_init(
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  event->impl->rmw_handle = rmw_get_zero_initialized_event();
+  event->impl->allocator = *allocator;
+
+  ret = rmw_publisher_event_init(
     &event->impl->rmw_handle,
     publisher->impl->rmw_handle,
     rmw_event_type);
+  if (ret != RMW_RET_OK) {
+    goto fail;
+  }
+
+  return RCL_RET_OK;
+fail:
+  allocator->deallocate(event->impl, allocator->state);
+  event->impl = NULL;
+  return ret;
```

Comment:
do we need to introduce `goto` statement here? instead of that i would do the following.

1. check error if not `RMW_RET_OK`.
2. if error happens, call deallocation.
3. then convert return code from `rmw` to `rcl`.
4. return it.

what do you think?

Reply:
I thought that before. I choose this one because I was thinking it will be less modification if there will be some extra logic in the future.

Reply:
okay understood.